### PR TITLE
Redux scaffolding

### PR DESCRIPTION
**Pull request by evmiguel**
This PR is for the initial redux scaffolding of the React application. The current scaffolding is integrated with the login flow.

### CODE REVIEW

File: client/components/UserSettings/index.jsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    username: PropTypes.string,
+    name: PropTypes.string
+};
+
+const mapStateToProps = state => {
+    const { isLoggedIn, username, name } = state.login;
+    return { isLoggedIn, username, name };
+};
+
+export default connect(mapStateToProps)(UserSettings);
```
Comment (spectranaut) [inline client/components/UserSettings/index.jsx:31]: So is this a line we will have to add to every page? How do we know what data from the data store a particular page will need?

File: client/reducers/index.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { combineReducers } from 'redux';
+import login from './login';
+
+export default combineReducers({
+    login
```
Comment (spectranaut) [inline client/reducers/index.js:5]: Oh, is the the entry point to all the data...

File: client/store/index.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { createStore } from 'redux';
+import rootReducer from '@client/reducers';
+import middleware from '@client/middleware';
+
+export default createStore(rootReducer, middleware);
```
Comment (spectranaut) [inline client/store/index.js:5]: I see. So there is one data store for the whole application. And then you choose what data you want in your stateToProps call?

File: server/services/GithubService.js [hunk 4]
```diff
@@ -16,5 +16,27 @@ module.exports = {
             }
         );
         return response.data.access_token;
+    },
+    async getUser(options) {
+        if (typeof options === 'undefined' || !('accessToken' in options)) {
+            throw new Error('Not authorized.');
+        }
+        const graphQLEndpoint = 'https://api.github.com/graphql';
```
Comment (spectranaut) [inline server/services/GithubService.js:24]: good thing you already know all about graphql :)

File: client/routes/index.js [hunk 5]
```diff
@@ -10,6 +11,11 @@ export default [
                 exact: true,
                 component: Login,
                 githubClientId: process.env.GITHUB_CLIENT_ID
+            },
+            {
+                path: '/account/settings',
```
Comment (spectranaut) [inline client/routes/index.js:16]: How does this control how things are displayed...

File: client/components/App/index.jsx [hunk 6]
```diff
@@ -1,18 +1,37 @@
...
+    }
+    render() {
+        const { route } = this.props;
+        return (
+            <div>
+                <h1>ARIA-AT REPORT</h1>
+                <Link to="/login">Login</Link>
+                <br></br>
+                <Link to="/account/settings">Settings</Link>
+                {renderRoutes(route.routes)}
```
Comment (spectranaut) [inline client/components/App/index.jsx:25]: So when a route is update, this is replace with what is specified in `client/routes/index.js`?

**Comment by spectranaut** [comment]
The failure is a lighthouse failure that I'll fix in a follow up PR. Merging this for now.

File: client/components/App/index.jsx [hunk 7]
```diff
@@ -1,18 +1,37 @@
...
+        if (login) {
+            dispatch(handleLogin());
+        }
+    }
+    render() {
+        const { route } = this.props;
+        return (
+            <div>
+                <h1>ARIA-AT REPORT</h1>
+                <Link to="/login">Login</Link>
```
Comment (rwaldron) [inline client/components/App/index.jsx:22]: What happens to this after the user has logged in? Also, `<Link to="/login">Login</Link>` should be `<Link to="/login">Log In</Link>` or `<Link to="/login">Log in</Link>` (https://github.com/bocoup/bocoup-meta/wiki/Style-Voice-and-Tone-Guide#notable-bocoup-or-in-the-field-style-points)

File: client/components/UserSettings/index.jsx [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+        return <div data-test="component-user-settings">{contents}</div>;
+    }
+}
+
+UserSettings.propTypes = {
+    isLoggedIn: PropTypes.bool,
+    username: PropTypes.string,
+    name: PropTypes.string
+};
```
Comment (rwaldron) [inline client/components/UserSettings/index.jsx:24]: Is there an `id` associated with each user?

File: client/index.js [hunk 9]
```diff
@@ -2,12 +2,16 @@ import React from 'react';
...
+import routes from './routes';
+import store from './store';
 
 ReactDOM.render(
-    <BrowserRouter>
-        <div>{renderRoutes(routes)}</div>
-    </BrowserRouter>,
+    <Provider store={store}>
+        <BrowserRouter>
+            <div>{renderRoutes(routes)}</div>
```
Comment (rwaldron) [inline client/index.js:13]: According to the readme examples for `react-router-dom`, the `<div>` wrapper is not necessary

File: client/reducers/login.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import { LOG_IN } from '../actions/types';
+
+const initialState = {
+    isLoggedIn: false
```
Comment (rwaldron) [inline client/reducers/login.js:4]:
```
username: null, 
name: null
```

File: client/reducers/login.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+};
+
+export default (state = initialState, action) => {
+    switch (action.type) {
+        case LOG_IN: {
+            const { username, name } = action.payload;
+            return {
+                isLoggedIn: true,
+                username,
+                name
```
Comment (rwaldron) [inline client/reducers/login.js:14]: Do you expect `payload` to have other irrelevant properties? If not, then you can do this:

```js
return {
  isLoggedIn: true,
  ...payload
};
```

This also guarantees that any new data that you add to the response from your API is automatically included

File: client/tests/App.test.jsx [hunk 12]
```diff
@@ -1,13 +1,23 @@
...
 Enzyme.configure({ adapter: new EnzymeAdapter() });
 
+const setup = (initialState = {}) => {
+    const store = storeFactory(initialState);
+    // Step into the higher order connected component and step into the contents of the UserSettings component
+    const wrapper = shallow(
+        <App route={{ routes }} store={store} location={{}} />
+    ).dive();
+    return wrapper;
+};
```
Comment (rwaldron) [inline client/tests/App.test.jsx:18]: Excellent!

File: client/tests/integration.test.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        moxios.wait(() => {
+            const request = moxios.requests.mostRecent();
+            request.respondWith({
+                status: 200,
+                response: {
+                    username: expectedState.login.username,
+                    name: expectedState.login.name
+                }
+            });
+        });
```
Comment (rwaldron) [inline client/tests/integration.test.js:33]: WDYT?

```js
const response = {
  username: 'foobar',
  name: 'Foo Bar'
};

const expectedState = {
  login: {
    isLoggedIn: true,
    ...response
  }
};

moxios.wait(() => {
  const request = moxios.requests.mostRecent();
  request.respondWith({
    status: 200,
    response
  });
});
```

**Pull request merged**